src/heartbeats.cpp

```diff
@@ -0,0 +1,605 @@
+    }
+
+    // handle pending updates
+    const auto &pendingSyncPair = mPendingSyncUpdates.find(syncTag);
+    if (pendingSyncPair != mPendingSyncUpdates.end())
+    {
+        if (!ISUNDEF(backupId))
+        {
+            updateSyncInfo(backupId, pendingSyncPair->second.get());
+        }
+        else
+        {
+            LOG_warn << "discarding heartbeat update for pending sync: no valid id received for sync: " << pendingSyncPair->second->getLocalFolder();
+
+        }
+        mPendingSyncUpdates.erase(pendingSyncPair);
+    }
+}
+
+int MegaBackupMonitor::getHBState(MegaSync *sync)
+{
+    if (sync->isTemporaryDisabled())
+    {
+        return MegaBackupMonitor::State::TEMPORARY_DISABLED;
+    }
+    else if (sync->isActive())
+    {
+        //TODO: consider use case: paused, if transfers are paused!?
+        return MegaBackupMonitor::State::ACTIVE;
+    }
+    else if (!sync->isEnabled())
+    {
+        return MegaBackupMonitor::State::DISABLED;
+    }
+    else
+    {
+        return MegaBackupMonitor::State::FAILED;
+    }
+}
+
+int MegaBackupMonitor::getHBSubstatus(MegaSync *sync)
+{
+    return sync->getError();
+}
+
+string MegaBackupMonitor::getHBExtraData(MegaSync *sync)
+{
+    return "";
+}
+
+BackupType MegaBackupMonitor::getHBType(MegaSync *sync)
+{
+    return BackupType::TWO_WAY; // TODO: get that from sync whenever others are supported
+}
+
+void MegaBackupMonitor::updateSyncInfo(handle backupId, MegaSync *sync)
+{
+
+    string localFolderEncrypted(mClient->cypherTLVTextWithMasterKey("lf", sync->getLocalFolder()) );
+    string deviceIDEncrypted(mClient->cypherTLVTextWithMasterKey("de", mClient->getDeviceid()) );
+    string nameEncrypted(mClient->cypherTLVTextWithMasterKey("na", sync->getName()) );
+
+
+    mClient->reqs.add(new CommandBackupPut(mClient, backupId, BackupType::TWO_WAY, sync->getMegaHandle(), localFolderEncrypted.c_str(),
+                                           deviceIDEncrypted.c_str(), nameEncrypted.c_str(),
+                                           getHBState(sync), getHBSubstatus(sync), getHBExtraData(sync).c_str()
+                                           ));
+}
+
+void MegaBackupMonitor::updateOrRegisterSync(MegaSync *sync)
+{
+    if (!sync)
+    {
+        return;
+    }
+    auto config = mClient->syncConfigs->get(sync->getTag());
+
+    handle syncID = UNDEF;
+    if (config)
+    {
+        syncID = config->getBackupId();
+    }
+
+    bool pushInPending = false;
+
+    string extraData;
+    if (syncID == UNDEF) //register
+    {
+
+        string localFolderEncrypted(mClient->cypherTLVTextWithMasterKey("lf", sync->getLocalFolder()) );
+        string deviceIDEncrypted(mClient->cypherTLVTextWithMasterKey("de", mClient->getDeviceid()) );
+        string nameEncrypted(mClient->cypherTLVTextWithMasterKey("na", sync->getName()) );
+
+
+        if (std::find(mPendingBackupPuts.begin(), mPendingBackupPuts.end(), sync->getTag()) == mPendingBackupPuts.end())
```

Comment:
I think we should not use the `syncTag` as the key for the pending API commands. Not here, and not for the pending updates (in `mPendingSyncUpdates`). A possible workaround is to use the command's tag. Another approach would be to maintain a queue of lambda functions that digest the response of those API commands.

Reply:
used a lambda function that holds the tag